### ISSUE TITLE
Please provide way to not use portals, or enforce permissions on portals

### ISSUE THREAD

**Issue by BillDietrich** [flatpak/flatpak#3977]
## Linux distribution and version
Ubuntu MATE 20.04
## Flatpak version
1.6.5
## Description of the problem
IMO Flatpak's permissions/portals system is confusing, non-standard, and not what many users want.  User action in a "portal" is deemed to grant permission that overrides the permissions set outside the image (via CLI or Flatseal).  But as a desktop user/admin, I don't care whether an attempted access by an app was stimulated by user action or some other way.  The permissions I set outside the app should control that app, period, no exceptions.

For example, I want to say "Firefox can access only these 4 directories, period.  I don't care what caused the access from Firefox, whether the user asked for it in a dialog or not, Firefox always can access only these 4 directories."

Please add some setting in the permission UI (CLI and/or Flatseal) to say "these permissions apply to the WHOLE app: portals, file chooser dialogs, user actions, internal app code, everything".  The current situation would be covered by leaving that setting off, and what I want would be covered by turning it on.  Thanks.

**Comment by Arxcis** [comment]
Hi @BillDietrich. I think what you are asking for, is a way to disable specific portals. That should be possible in Flatpak already.

_Example of disabling portal.FileChooser:_
```sh
flatpak run org.mozilla.firefox --no-talk-name="org.freedesktop.potal.FileChooser"
```

This would solve your problem, wouldn't it?

**Comment by Arxcis** [comment]
**Update/Bug-report:** Actually, I tried running this command with `org.gnome.Eiphany`, which only has access to `xdg-download`-folder, and nothing else by default:
```
flatpak run --no-talk-name="org.freedesktop.portal.FileChooser" org.gnome.Epiphany
```
Filesystem permissions in `Flatseal`:
![image](https://user-images.githubusercontent.com/3092618/100137280-8c5aeb80-2e8c-11eb-9734-8d6156de73a2.png)


**Expected behaviour**
When I try to open a file, I expect the Epihany to not have access to the `org.freedesktop.portal.FileChooser`, and fall back to regular file-chooser, only seeing the `Download`-folder as an option. Here is a screenshot from Firefox displaying expected behaviour:
<img src="https://user-images.githubusercontent.com/3092618/100136531-74cf3300-2e8b-11eb-9919-8a2c7c74c7c8.png" width=500>

**Actual behaviour**
All files are still visible in the file-chooser, as if I still have access to the portal:
<img src="https://user-images.githubusercontent.com/3092618/100136670-9e885a00-2e8b-11eb-9eb3-3ddb0c202515.png" width=500>

**Versions**
```sh
$ uname -a && flatpak --version
Linux pop-os 5.8.0-7630-generic #32~1605108806~20.10~7e52b13-Ubuntu SMP Wed Nov 11 19:10:30 UTC  x86_64 x86_64 x86_64 GNU/Linux
Flatpak 1.8.2

$ flatpak info org.gnome.Epiphany

GNOME Web - Web browser for GNOME

          ID: org.gnome.Epiphany
         Ref: app/org.gnome.Epiphany/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: 3.38.0

$ flatpak info org.mozilla.firefox

Firefox - Mozilla Firefox Web Browser

          ID: org.mozilla.firefox
         Ref: app/org.mozilla.firefox/x86_64/beta
        Arch: x86_64
      Branch: beta
     Version: 84.0b3
```

**Comment by Arxcis** [comment]
Before I create a bug-ticket, could someone confirm that this actually a bug?

**Comment by Arxcis** [comment]
## Update update
If found something that worked-ish.
```
flatpak run --no-documents-portal org.gnome.Epiphany
```
 File-chooser still displays all files, but when any file is selected outside of `xdg-download` Epiphany is no longer able to load the file anymore.

**Comment by Arxcis** [comment]
**The root of the problem??**
When I add an override like **--no-talk-name**, it does not actually add a line to the override file. It just removes the line if it is already there. This leaves no room to express, that you want to turn **off** something that is by  **on** by default, like portals.

_Example:_
```sh
$ flatpak override --user --talk-name=org.freedesktop.portal.FileChooser org.gnome.Epiphany

$ flatpak info org.gnome.Epiphany --show-permissions
[Context]
shared=network;
sockets=wayland;pulseaudio;
devices=dri;
filesystems=xdg-download;/run/.heim_org.h5l.kcm-socket;

[Session Bus Policy]
org.freedesktop.portal.FileChooser=talk                    <------------ JUST ADDED

[System Bus Policy]
org.freedesktop.GeoClue2=talk

$ flatpak override --user --no-talk-name=org.freedesktop.portal.FileChooser org.gnome.Epiphany

$ flatpak info org.gnome.Epiphany --show-permissions
[Context]
shared=network;
sockets=wayland;pulseaudio;
devices=dri;
filesystems=xdg-download;/run/.heim_org.h5l.kcm-socket;

                                                                   <------------------ REMOVED
[System Bus Policy]
org.freedesktop.GeoClue2=talk
```

**What I really wanted to happen:**
```
$ flatpak info org.gnome.Epiphany --show-permissions
[Context]
shared=network;
sockets=wayland;pulseaudio;
devices=dri;
filesystems=xdg-download;/run/.heim_org.h5l.kcm-socket;

[Session Bus Policy]
org.freedesktop.portal.FileChooser=no-talk                    <------------ Updated

[System Bus Policy]
org.freedesktop.GeoClue2=talk
```

**Comment by Arxcis** [comment]
It is really wired that this happens, because the `test-override.sh` should have caught this bug:
https://github.com/flatpak/flatpak/blob/b1fdf6c4ef9d10b90a5d1004963f9c13606e2c15/tests/test-override.sh#L94-L103

**Comment by alexlarsson** [comment]
I wouldn't say it was designed to not let you turn off portals. But it wasn't designed to let you turn them off so it accidentally is not possible. Making it possible to be turned off would be fine with me, but its not really a personal priority for me.

**Comment by Arxcis** [comment]
@alexlarsson Thank you for clarifying. It makes sense to me that it is accidentally not possible :+1: I guess this issue is a feature-request then. Not a high priority for me anyway, but I can see that it may be useful in certain cases elaborated by OP @BillDietrich.

**Comment by JJRcop** [comment]
I'd like to clarify here to @BillDietrich that Flatpak and Flatseal are separate projects with separate developers. The way this would happen is first Flatpak would implement the overrides feature for disabling portals, then only after this could Flatseal expose UI to change the override.

But as stated, these are separate projects, so it is incorrect to refer to them as if they were different components of a single project.

**Comment by ssokolow** [comment]
I'll also weigh in on the necessity of this as, currently, Kubuntu Linux 20.04 LTS includes a version of `xdg-desktop-portal-kde` that doesn't support directory pickers (it just shows a file picker that doesn't let you pick anything) and the PPA only offers updates for `xdg-desktop-portal-gtk`.

Normally, I `export GTK_USE_PORTAL=1` to get KDE dialogs in even the non-Flatpak'd applications but, in this case, use of the portal renders the ScummVM Flatpak completely incapable of adding games. (In a moment, I'll be opening a bug report on it asking for them to either find a workaround or forcibly disable use of the portal on their end.)

(The only reason I can play the games I already have in it is because I copied an existing `scummvm.ini` from the older `apt` version which supports fewer games.)

If it were possible to disable the portal, then it would Just Work™ because I've manually granted `:ro` access to the folder where I keep my ScummVM game assets.

**Comment by eike-fokken** [comment]
I would like to second the feature request here. I like to severely lock down my flatpak-sandboxed apps.

If I, the user, can grant the read permission in the filechooser portal, probably a malicious program can as well, no?

**Comment by rusty-snake** [comment]
> probably a malicious program can as well, no?

A malicious flatpak or malicious program installed via e.g. dnf? If you mean a malicious program on your system which runs unsandboxed, this is nothing you need to worry about.

> org.freedesktop.portal.FileChooser=none 

`org.freedesktop.portal.FileChooser` is an interface and not an bus name.

> Is there any other, however hackish way to disallow the use of the filechooser?

Maybe `FLATPAK_DBUSPROXY=<wrapper script> flatpak run …`?

**Comment by eike-fokken** [comment]
> If you mean a malicious program on your system which runs unsandboxed, this is nothing you need to worry about.

What I really meant is this: Which pieces of software do I have to trust to be sure that I really need user intervention for the filechooser to grant permission to a file.

> org.freedesktop.portal.FileChooser is an interface and not an bus name.

I simply copied from @Arxcis s answer above. Unfortunately I don't know much about the settings of flatpak up to now. I always used Flatseal for setting permissions.

> Maybe FLATPAK_DBUSPROXY=<wrapper script> flatpak run …?

Could you elaborate on that? Apparently that environment variable is used in the tests but other than that I can't find much on it.

**Comment by smcv** [comment]
> If I, the user, can grant the read permission in the filechooser portal, probably a malicious program can as well, no?

No, a malicious program inside a Flatpak sandbox cannot use the portal to grant itself read permission on arbitrary files outside the sandbox.

The "Open" or "Save As" dialog is presented by a separate program *outside* the sandbox ([xdg-desktop-portal-gtk](https://github.com/flatpak/xdg-desktop-portal-gtk/), or some other portal backend like the `-gnome` or `-kde` implementations). Programs outside the sandbox are not protected by Flatpak, and can read and write all your files already.

After you select a file to open, the [xdg-desktop-portal](https://github.com/flatpak/xdg-desktop-portal/) implementation makes that file appear in a FUSE filesystem *inside* the sandbox (with read or read/write permissions, depending which file-chooser you were interacting with), so that the less-trusted program can access that file (but not others).

Making this happen transparently, without having to use something like Flatseal, is the entire point of the file-chooser portal (which is responsible for the "Open" or "Save As" dialog), and the documents portal (which is responsible for the FUSE filesystem).

> Which pieces of software do I have to trust to be sure that I really need user intervention for the filechooser to grant permission to a file

You need to trust anything that runs outside the Flatpak sandbox, in particular:

* Flatpak itself
* xdg-desktop-portal
* Whatever xdg-desktop-portal backend you are using, for example xdg-desktop-portal-gtk
* the Linux kernel

For completeness: if you have a Flatpak app installed that has been given permissions that would allow it to execute arbitrary code outside its sandbox, then you also need to trust that app. These permissions are the ones you manage with tools like Flatseal and `flatpak override`. Typically, IDEs like GNOME Builder have these elevated permissions, because they would not be useful otherwise. Ordinary apps like GNOME Web (Epiphany) do not.

**Comment by smcv** [comment]
> should the lines
> ```
> [Session Bus Policy]
> org.freedesktop.portal.FileChooser=none 
> ```
>
> have the effect of disallowing the filechooser dialogue?

No, for two reasons:

1. D-Bus permissions like `talk` only act on bus names, not interface names. Bus names are "larger" than interface names. `org.freedesktop.portal.FileChooser` is one of several interfaces implemented by `xdg-desktop-portal`. Its bus name is `org.freedesktop.portal.Desktop`.
2. There is a special-case for access to bus names matching `org.freedesktop.portal.*`, which are always allowed. Most non-portal bus names like `com.example.One` or `org.freedesktop.Two` are only allowed if there is an explicit `talk` permission. However, the portal names are always allowed, because the purpose of a portal is that it takes responsibility for doing appropriate permissions handling (either from configuration, or by prompting the user interactively, or whatever) before taking privileged actions outside the sandbox on a sandboxed app's behalf.

(For completeness: there is also special handling for `org.freedesktop.DBus`, which needs to be *partially* allowed by default, otherwise D-Bus communication would not work at all.)

**Comment by eike-fokken** [comment]
Ok, sounds reasonable. Although if I understand correctly, more security could be gained if I could choose to not use a portal and instead make the file dialog of an app also honor the permissions set in the flatpak config file. In that case I would only need to trust flatpak and its sandbox implementation and of course the kernel.

I must say, this is not paranoid enough for my taste, but I also understand that flatpak chooses different trade-offs.

**Comment by thomasdevriese** [comment]
Hi, I'm running Chromium inside Flatpak, and one of the requirements is to prevent the user from downloading and uploading files (just seeing the file structure of the host system is also strictly prohibited). My idea was to disable the filechooser portal, but if I read this thread correctly, this would not be possible? Does someone know if there are any other ways to achieve this?

I also thought of setting the [disable-save-to-disk property](https://docs.flatpak.org/en/latest/portal-api-reference.html#gdbus-property-org-freedesktop-impl-portal-Lockdown.disable-save-to-disk) to true, but it seems I can't communicate with the _org.freedesktop.impl.portal.Lockdown_ interface as it is not under the _org.freedesktop.portal.Desktop_ bus.

Thanks!

**Comment by orowith2os** [comment]
@smcv even blocking org.freedesktop.portal.Desktop doesn't work to disable filechooser portal access, it seems, @TheEvilSkeleton is having issues with disabling it to fix an issue with Krita (https://bugs.kde.org/show_bug.cgi?id=438608). Does Flatpak hard-code the permissions for these interfaces still?

**Comment by rusty-snake** [comment]
> There is a special-case for access to ***bus names matching `org.freedesktop.portal.*`,  which are always allowed.***

**Comment by prahal** [comment]
flatpak does not really give a way to set file permission per application but per sandbox (even if most flatpak sandboxes are holding only one application).

To me, the wish of the OP is to control file access per application. Neither flaptak nor portal is meant to provide this feature.

The OP asking for flatpak file permission model to control portal too is not required to achieve his goal.
Looks to me this is the job for AppArmor or SELinux, though I don't know if one can add filters for these for flatpak packaged applications.

**Comment by Mikenux** [comment]
Perhaps it is necessary to determine if Flatpak requires changes so that any tools limiting file access can limit it to Flatpak apps?

Additionally, reading the use case of disallowing downloading and uploading of files (https://github.com/flatpak/flatpak/issues/3977#issuecomment-1308967577). I think it makes sense to disable the display of the file chooser. Indeed (and if this is not already the case), it would make no sense to present users with a file chooser if they are not allowed to access any files.

**Comment by ssokolow** [comment]
Fundamentally, Flatpak's design philosophy is built around "If a user asks for something and gets told they can't do it, they'll blame Flatpak and seek to replace it with another distribution channel".

Flatpak isn't a MAC system like SELinux. It's about constraining the programs to what the user asks for, not about constraining the user to some kind of system policy.

...so any constraints on what the user can do must very visibly *not* be tied to the presence or absence of Flatpak. (Which is a good idea anyway, because tying system policy constraints too strongly to Flatpak just gives you a false sense of security in the face of confused deputy exploits.)

**Comment by alidehghan** [comment]
After two years, I verify that this problem still exists and there is no way to control the FileChooser dialog. That is not cool.

**Comment by arnosec** [comment]
this problem exists. I tried everything, info shows that librewolf only has access to xdg-download but it has access to home dir and not only downloads. this is a serious issue and its been here for long time now, im looking at alternative to flatpak because it's simply not secure at this point. flatpak is designed to make apps  available and run the without dependency issues , separated from the system , not for sandboxing apps. i was using it because i thought i would sandbox also by only allowing rw access to xdg-download only for browser and other similar apps and nothing else other than .var dir. but that's simply not the case. overriding simply is removing a line from info permissions when you view it, and in actual browser still has access to home dir

**Comment by ssokolow** [comment]
@arnosec Have you confirmed that `flatpak run --command=bash io.gitlab.librewolf-community` can see your home directory? ...because that's `bash` running under the same conditions Librewolf sees.

**Comment by razzeee** [comment]
You will also need to restart the app after permission changes. Could also be your distro misconfiguring flatpak.

**Comment by swick** [comment]
I don't think there is anything useful coming out of this issue anymore. If you run up-to-date software and experience a bug, open a new issue.

**Comment by beedell-roke** [comment]
@swick, what has changed?

**Comment by arnosec** [comment]
> You will also need to restart the app after permission changes. Could also be your distro misconfiguring flatpak.

as i said before, i tried everything i can. i even tested it in debian,arch,artix and void linux. same result everywhere. flatpak is not designed to enforce policies and simply not a sandboxing app.

**Comment by RundownRhino** [comment]
The recent discussion in this thread seems to be on a different topic than the original issue, that there's no way to disable specific portals. That issue is still present.

One example of why I think this is an important feature is that currently, any app launched in a flatpak without network access can still exfiltrate data through a GET request, by putting it into a link's querystring and opening it up in a browser via the OpenURI portal. There is currently no way to prevent this, except by disabling the `flatpak-portal` service and breaking *all* portals.

**Comment by voidpointertonull** [comment]
> The recent discussion in this thread seems to be on a different topic than the original issue, that there's no way to disable specific portals. That issue is still present.

Can confirm this issue.
Personally I was "thrilled" to learn about a portal broadcasting the lock state change of the desktop environment after an update got it working on my system, so Telegram's anti-feature of creating useless notifications while the host is locked appeared. Found no way to disable the portal, found no way to restrict permissions, it was just simply decided that all Flatpak programs are entitled to know about when I'm leaving my desktop and when I'm coming back, even if I'm not interacting with them at all.

This issue getting hijacked is a good example of the mess we have, people don't even know where's information leaking from, or if it's not actually leaking just not presented well due to not having a clearly marked boundary.
The leak could be from a portal, from a D-Bus connection, or maybe even from an explicit container hole like in #4311 , it's hard to track down which part of the swiss cheese is responsible for a specific problem.

> One example of why I think this is an important feature is that currently, any app launched in a flatpak without network access can still exfiltrate data through a GET request, by putting it into a link's querystring and opening it up in a browser via the OpenURI portal.

With how common that is, unfortunately that's specifically a web permissiveness problem with quite a few programs having the assumption already of being able to just open URLs to nag the user.

However it's definitely silly how easy it is to cross container boundaries not with just URLs, but even with files.
If an exploit for my PDF reader would get discovered, I wouldn't be able to do anything else than just removing the PDF reader as the default PDF handler, even if I'd still trust all the PDFs downloaded earlier especially in a different security context.
